contents/kotlin/NullSafety.md

```diff
@@ -0,0 +1,121 @@
+# Null Safety in Kotlin
+
+Author: [Pan Haozhe](https://github.com/Haozhe321)
+
+# Overview
+Another part of this book chapter gives an introduction to Kotlin. This piece aims to discuss the null safety feature in Kotlin, and what was previously missing in Java.
+
+# The Billion Dollar Mistake
+“I call it my billion-dollar mistake. It was the invention of the null reference…My goal was to ensure that all use of references should be absolutely safe, with **checking performed automatically by the compiler**. But I couldn't resist the temptation to put in a null reference, simply because it was so easy to implement.”  
+-Tony Hoare  
+
+
+# NullPointerException
+As we know, Kotlin is now supported as the official language of Android applications. When developing Android applications in Java, NullPointerException(NPE) was a big problem. In fact, About one third of app crashes can be attributed to NPE. To see how it happens, let's take a look at the piece of Java code below:
+
+```java
+String a = null;
+if(a.length > 5) {
+    //do something
+}
+```
+When the above code is ran, a NPE will be thrown on line 2 because a `null` object has no methods. To prevent an object from taking on a `null` value, programmers resort to doing additional checks like this:
+```java
+String a = null;
+if(a != null) {
```

Comment:
This example isn't so good because it has been long agreed that preventing nulls by default is a good thing. See how Java 7 introduced `Objects#requireNonNull(Object)`.

On the other hand, if this value needs to be optional, it is not a good idea to conflate nulls and optional values.

For instance, for pre-Java 8 there's Guava `Optional` and for Java 8 onwards there's also `Optional`.

Then we can do `a.ifPresent(this::doSomething);`

Also, a more readable (common) way of writing a check like this would not be so much nesting:
```java
String a = null;
if(a != null &&a.length > 5) {
    //do something
}
```